File: app/src/main/java/forpdateam/ru/forpda/ui/fragments/devdb/device/specs/SpecsAdapter.java
package forpdateam.ru.forpda.ui.fragments.devdb.device.specs;

import android.support.v7.widget.RecyclerView;
import android.util.Pair;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

import forpdateam.ru.forpda.App;
import forpdateam.ru.forpda.R;
import forpdateam.ru.forpda.model.data.remote.api.ApiUtils;

/**
 * Created by radiationx on 08.08.17.
 */

public class SpecsAdapter extends RecyclerView.Adapter<SpecsAdapter.ViewHolder> {
    private ArrayList<Pair<String, List<Pair<String, String>>>> list = new ArrayList<>();


    public void addAll(Collection<Pair<String, List<Pair<String, String>>>> results) {
        addAll(results, true);
    }

    public void addAll(Collection<Pair<String, List<Pair<String, String>>>> results, boolean clearList) {
        if (clearList)
            clear();
        list.addAll(results);
        notifyDataSetChanged();
    }

    public void clear() {
        list.clear();
    }

    @Override
    public SpecsAdapter.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.device_spec_item, parent, false);
        return new SpecsAdapter.ViewHolder(v);
    }

    @Override
    public void onBindViewHolder(SpecsAdapter.ViewHolder holder, int position) {
        Pair<String, List<Pair<String, String>>> item = list.get(position);
        holder.title.setText(item.first);
        StringBuilder builder = new StringBuilder();

        for (int i = 0; i < item.second.size(); i++) {
            Pair<String, String> pair = item.second.get(i);
            String strColor = String.format("#%06X", 0xFFFFFF & App.getColorFromAttr(holder.itemView.getContext(), R.attr.second_text_color));
            builder.append("<small style=\"font-size:10px\"><span style=\"color: ").append(strColor).append("\">").append(pair.first).append("</span></small><br>").append(pair.second);
            if (i + 1 < item.second.size()) {
                builder.append("<br><br>");
            }
        }

        holder.desc.setText(ApiUtils.coloredFromHtml(builder.toString()));
        /*holder.price.setVisibility(item.getPrice() == null ? View.GONE : View.VISIBLE);
        if (item.getPrice() != null) {
            holder.price.setText(item.getPrice());
        }*/

    }

    @Override
    public int getItemCount() {
        return list.size();
    }

    public Pair<String, List<Pair<String, String>>> getItem(int position) {
        return list.get(position);
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        public TextView title;
        public TextView desc;

        public ViewHolder(View v) {
            super(v);
            title = (TextView) v.findViewById(R.id.item_title);
            desc = (TextView) v.findViewById(R.id.item_desc);
        }

    }
}

File: app/src/main/java/forpdateam/ru/forpda/ui/fragments/devdb/device/specs/SpecsAdapter.kt
package forpdateam.ru.forpda.ui.fragments.devdb.device.specs

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import by.kirich1409.viewbindingdelegate.viewBinding
import forpdateam.ru.forpda.R
import forpdateam.ru.forpda.databinding.DeviceSpecItemBinding
import forpdateam.ru.forpda.extensions.getColorFromAttr
import forpdateam.ru.forpda.model.data.remote.api.ApiUtils

/**
 * Created by radiationx on 08.08.17.
 */
class SpecsAdapter : RecyclerView.Adapter<SpecsAdapter.ViewHolder>() {
    private val list = ArrayList<Pair<String, List<Pair<String, String>>>>()


    @JvmOverloads
    fun addAll(
        results: Collection<Pair<String, List<Pair<String, String>>>>,
        clearList: Boolean = true
    ) {
        if (clearList) clear()
        list.addAll(results)
        notifyDataSetChanged()
    }

    fun clear() {
        list.clear()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val v =
            LayoutInflater.from(parent.context).inflate(R.layout.device_spec_item, parent, false)
        return ViewHolder(v)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(getItem(position))
    }

    override fun getItemCount(): Int {
        return list.size
    }

    fun getItem(position: Int): Pair<String, List<Pair<String, String>>> {
        return list[position]
    }

    class ViewHolder(v: View) : RecyclerView.ViewHolder(v) {

        private val binding by viewBinding<DeviceSpecItemBinding>()

        fun bind(item: Pair<String, List<Pair<String, String>>>) {
            binding.itemTitle.text = item.first
            val builder = StringBuilder()

            for (i in item.second.indices) {
                val pair = item.second[i]
                val strColor = String.format(
                    "#%06X",
                    0xFFFFFF and binding.root.context.getColorFromAttr(R.attr.second_text_color)
                )
                builder.append("<small style=\"font-size:10px\"><span style=\"color: ")
                    .append(strColor)
                    .append("\">").append(pair.first).append("</span></small><br>")
                    .append(pair.second)
                if (i + 1 < item.second.size) {
                    builder.append("<br><br>")
                }
            }

            binding.itemDesc.text = ApiUtils.coloredFromHtml(builder.toString())
        }
    }
}
